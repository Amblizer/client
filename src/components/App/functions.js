import io from 'socket.io-client';

const remoteHost = process.env.REMOTE_HOST || 'localhost';
const remotePort = process.env.REMOTE_PORT || '8000';

const socketPath = `http://${remoteHost}:${remotePort}`;

export function createSocket (app) {
  const socket = io.connect(socketPath, {
    secure: false,
    reconnectionAttempts: 10,
    query: 'type=client',
  });

  socket.on('operator:message', app.receiveMessage);
  socket.on('chat:new', app.handleNewConnection);

  return socket;
}

export function canCombineLastMessage (msg, messages) {
  const lastMsg = messages[messages.length - 1];

  // If this is the first message in the conversation
  if (lastMsg === undefined) return false;

  // If theres no author field, return false
  if (!msg.hasOwnProperty('author') || !lastMsg.hasOwnProperty('author')) return false;

  // If last message was not from the same author
  if (msg.author !== lastMsg.author) return false;

  return true;
}

export function combineLastMessage (msg, messages) {
  const lastMsg = messages[messages.length - 1];
  const newMsg = msg;

  newMsg.content = newMsg.content || [];

  // Is the content of the new message an array?
  if (!(newMsg.content.push instanceof Function)) {
    newMsg.content = [newMsg.content];
  }

  // Can we combine the last message?
  if (canCombineLastMessage(msg, messages)) {
    const combinedMessages = [...messages];

    // If there is no content in the last message make sure we can add some
    lastMsg.content = lastMsg.content || [];

    // Is the content of the last message an array?
    if (!(lastMsg.content.push instanceof Function)) {
      lastMsg.content = [lastMsg.content];
    }

    // Push contents to last message
    lastMsg.content.push(...newMsg.content);

    // Splice in new last message
    // TODO: This may not be necessary if lastMsg is a reference to messages?
    combinedMessages.splice(combinedMessages.length - 1, 1, lastMsg);

    return combinedMessages;
  }

  return [...messages, newMsg];
}
