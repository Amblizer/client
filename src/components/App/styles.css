/* Application Component */

@import url('https://fonts.googleapis.com/css?family=Open+Sans:400,700');

#mnml-chat {
  font-size: 12px;
  font-family: 'Open Sans', Arial;
}


/* Messenger Theme */

.mnml--messenger {
  position: fixed;
  display: flex;
  flex-direction: column;
  bottom: 0;
  right: 100px;
  width: 280px;
  height: 360px;
}

.mnml--messenger.closed {
  position: fixed;
  display: flex;
  flex-direction: column;
  bottom: 0;
  right: 100px;
  width: 280px;
  height: 32px;
}

/* Float Theme */

