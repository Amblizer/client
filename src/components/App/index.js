import { h, Component } from 'preact';

import Chat from '../Chat';
import ClosedState from '../ClosedState';
import { ThemeProvider } from '../ThemeProvider';
import {
  formatMessageForClient,
  formatMessageForServer,
  combineLastMessage,
  createSocket,
} from './functions';

import './styles.css';

const MESSENGER = 'messenger';
const FLOAT = 'float';
const SIDEPANEL = 'side';

class App extends Component {
  state = {
    chatOpen: false,
    messages: [],
    textBox: '',
    theme: MESSENGER, // wrapped with theme provider + HOC
  };

  componentDidMount () {
    this.socket = createSocket(this);
  }

  // Event Handlers

  toggleChat = bool => {
    this.setState({ chatOpen: bool });
  };

  handleInput = e => {
    this.setState({ textBox: e.target.value });
  };

  //  Socket Methods

  /**
  * @description On connecting to the socket server save a session object into the state
  */
  handleNewConnection = e => {
    this.setState({
      session: JSON.parse(e),
    });
  };

  //  Message Methods

  // TODO: docstring
  // save the message to local stage: either combining them or not.
  saveMessageToState = msg => {
    const formattedMsg = formatMessageForClient(
      msg,
      this.state.session.client.id,
      this.state.session.id
    );

    this.setState({
      messages: combineLastMessage(formattedMsg, this.state.messages),
      textBox: '',
    });
  };

  saveMessageToServer = msg => {
    const formattedMsg = formatMessageForServer(
      msg,
      this.state.session.client.id,
      this.state.session.id
    );

    this.socket.emit('client:message', JSON.stringify(formattedMsg));
  };

  /** Send Message
   * @summary - Allow a user to send a message.
   * @description - Passes the message through some functions to save it locally and remotely
   * @param {object} e - event object; used to prevent refreshing the page
   */
  sendMessage = e => {
    e.preventDefault(); // Must prevent default behavior first

    if (this.state.textBox === '') return;

    const msg = this.state.textBox;

    this.saveMessageToState(msg);

    this.saveMessageToServer(msg);
  };

  receiveMessage = data => {
    const msg = JSON.parse(data); // Data comes in as a string

    this.setState({
      messages: combineLastMessage(msg, this.state.messages),
    });
  };

  // Render
  //

  renderClosedChat = () => (
    <ClosedState
      chatOpen={this.state.chatOpen}
      toggleChat={this.toggleChat}
    />
  )

  renderOpenChat = () =>
    (<Chat
      toggleChat={this.toggleChat}
      messages={this.state.messages}
      textBox={this.state.textBox}
      handleInput={this.handleInput}
      sendMessage={this.sendMessage}
      chatOpen={this.state.chatOpen}
    />);

  renderChat = () => (this.state.chatOpen ? this.renderOpenChat() : this.renderClosedChat());

  render () {
    const { theme, chatOpen } = this.state;
    const visibility = chatOpen ? 'open' : 'closed';

    return (
      <ThemeProvider theme={this.state.theme}>
        <div className={`mnml--${theme} ${visibility}`}>
          {this.renderChat()}
        </div>
      </ThemeProvider>
    );
  }
}

export default App;
