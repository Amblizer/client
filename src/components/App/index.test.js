import { h } from 'preact';
import render from 'preact-render-to-string';
import io from 'socket.io-client';

import App from '.';

describe('<App />', () => {
  it('matches snapshot', () => {
    const tree = render(<App />);
    expect(tree).toMatchSnapshot();
  });
});
