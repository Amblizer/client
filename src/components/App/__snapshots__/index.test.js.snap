// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> matches snapshot 1`] = `"<div class=\\"mnml--messenger open\\"><section class=\\"Chat--messenger\\"><header class=\\"Header--messenger\\"><span class=\\"Header__title--messenger\\">Chat with John</span><button class=\\"Header__closeBtn--messenger\\">Ã—</button></header><ul class=\\"Chat__body--messenger\\"></ul><form class=\\"Input__form--messenger\\"><input class=\\"Input--messenger\\" placeholder=\\"Type Here\\" name=\\"messages\\" value /></form></section></div>"`;
